### PR TITLE
[WIP] Reimplement per player view distance API

### PR DESCRIPTION
**Pull request by kickash32**
This API was broken when paper updated to 1.14 and was replaced with a stub. This PR removes the stub API and reimplements the original patch. The patch is largely unchanged outside of the PlayerChunkMap class and PlayerList class. 

All instances of the viewDistance field from PlayerChunkMap and viewDistance field from SpigotWorldConfig, were changed to use a player's view distance wherever applicable.

Best View Distance view distance was used to test the API and it seems to work as expected.
image

I believe further testing is required.

### CODE REVIEW

**Comment by Spottedleaf** [comment]
Where is the code for player tickets?

**Comment by kickash32** [comment]
I am unsure if it is required as it is missing from the vanilla code unlike 1.13.

**Comment by Spottedleaf** [comment]
You'll find that it is required if you attempt to go over the world view distance.

**Comment by kickash32** [comment]
Did some testing, and that doesn't seem to be the case. Tested with player view distance at 12, world view distance at 3, server view distance at 10.

**Comment by Spottedleaf** [comment]
I don't think you've tested correctly then. Chunks cannot be sent to players if they're not at ticking status, and when you do not add player tickets to chunks they do not get ticking status. There's also other concerns along that line.

File: Spigot-Server-Patches/0409-Add-player-view-distance-API.patch [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+-                this.a(chunkcoordintpair, false).forEach((entityplayer) -> {
+-                    int i1 = b(chunkcoordintpair, entityplayer, true);
+-                    boolean flag = i1 <= finall; // CraftBukkit - decompile error
+-                    boolean flag1 = i1 <= this.viewDistance;
+-
+-                    this.sendChunk(entityplayer, chunkcoordintpair, apacket, flag, flag1);
+-                });
++            // Paper start - Separate into two methods
++            if(this.world.players != null) {
++                new ArrayList<>(this.world.players).forEach((player) -> this.setViewDistance(player, k));
```
Comment (kamcio96) [inline Spigot-Server-Patches/0409-Add-player-view-distance-API.patch:83]: Why `new ArrayList` ?

Comment (Ghost-chu) [reply Spigot-Server-Patches/0409-Add-player-view-distance-API.patch:83]: i think that can allow use lambda to foreach

**Comment by kickash32** [comment]
Just so everyone is aware, this patch is incomplete due to issues as highlighted by @Spottedleaf. The patch needs to adjust how tickets are added so that they respect each player's view distance. This is unfortunately very difficult and I have given up for the time being.

File: Spigot-Server-Patches/0409-Add-player-view-distance-API.patch (hunk 1 above)
Comment (KernelFreeze) [reply Spigot-Server-Patches/0409-Add-player-view-distance-API.patch:83]: Or you can use `Arrays.stream(this.world.players)`, so Java doesn't need to allocate new memory 😄

**Pull request closed without merging**